### PR TITLE
Add tests for trace validation, cleanup validations

### PR DESCRIPTION
**Pull request by the-ericwang35**
Adds unit tests that test errors for invalid protos, move `CheckForRequiredFields` to a shared file, add header to some files

### CODE REVIEW

File: validation/mock_metric_validation.go [hunk 1]
```diff
@@ -1,15 +1,27 @@
...
-	"fmt"
 	"reflect"
 
 	"google.golang.org/genproto/googleapis/monitoring/v3"
-	"google.golang.org/genproto/googleapis/rpc/errdetails"
 )
 
 const (
-	missingFieldMsg = "%v must contain requried %v field"
+	missingFieldMsg = "%v must contain required %v field"
```
Comment (mabdi3) [inline validation/mock_metric_validation.go]: We can move this field into the shared file as well.

Comment (the-ericwang35) [reply validation/mock_metric_validation.go]: good point, done

File: server/trace/mock_trace_test.go [hunk 2]
```diff
@@ -129,6 +162,63 @@ func TestMockTraceServer_BatchWriteSpans(t *testing.T) {
 	}
 }
 
+func TestMockTraceServer_BatchWriteSpans_Error(t *testing.T) {
```
Comment (mabdi3) [inline server/trace/mock_trace_test.go]: Maybe split this up for testing invalid timestamps and missing fields, it's a little packed.

**Review by mabdi3 — APPROVED** [review]

File: server/trace/mock_trace_test.go (hunk 2 above)
Comment (the-ericwang35) [reply server/trace/mock_trace_test.go]: done

File: validation/mock_trace_validation.go [hunk 3]
```diff
@@ -1,24 +1,37 @@
...
-
-	"github.com/golang/protobuf/ptypes"
 )
 
 var (
-	InvalidTimestampErr   = status.Error(codes.InvalidArgument, "start time must be before end time")
-	MalformedTimestampErr = status.Error(codes.InvalidArgument, "unable to parse timestamp")
-	MissingFieldError     = status.New(codes.InvalidArgument, "Missing required field(s)")
-	requiredFields        = []string{"Name", "SpanId", "DisplayName", "StartTime", "EndTime"}
+	InvalidTimestampError   = status.Error(codes.InvalidArgument, "start time must be before end time")
```
Comment (tbarker25) [inline validation/mock_trace_validation.go]: I think the convention for Go is typically
```
ErrInvalidTimestamp = ...
ErrMalformedTimstamp = ...
```

See https://golang.org/pkg/io/#pkg-variables

File: validation/shared_validation.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+const (
+	missingFieldMsg = "%v must contain required %v field"
+)
+
+func CheckForRequiredFields(requiredFields []string, reqReflect reflect.Value, requestName string) error {
+	br := &errdetails.BadRequest{}
+
+	for _, field := range requiredFields {
+		if isZero := reflect.Indirect(reqReflect).FieldByName(field).IsZero(); isZero {
```
Comment (tbarker25) [inline validation/shared_validation.go]: `if reflect.Indirect(reqReflect).FieldByName(field).IsZero() {`

File: server/trace/mock_trace_test.go [hunk 5]
```diff
@@ -97,55 +102,195 @@ func generateSpan(spanName string) *cloudtrace.Span {
...
+		}
+	}
+}
 
-		if !proto.Equal(responseSpan, c.want) {
-			t.Errorf("CreateSpan(%q) == %q, want %q", c.in, responseSpan, c.want)
+func validateErrDetails(err error, missingFields map[string]struct{}) bool {
+	st := status.Convert(err)
+	for _, detail := range st.Details() {
+		switch t := detail.(type) {
```
Comment (tbarker25) [inline server/trace/mock_trace_test.go]: probably clearer as an `if` since its just a single case:
```
if t, ok := detail.(type); ok {
   ...
}
```

**Review by tbarker25 — APPROVED** [review]

File: validation/mock_trace_validation.go (hunk 3 above)
Comment (the-ericwang35) [reply validation/mock_trace_validation.go]: Got it, done

File: server/trace/mock_trace_test.go (hunk 5 above)
Comment (the-ericwang35) [reply server/trace/mock_trace_test.go]: Thanks, done

**Pull request merged**